test/core/util/test_lb_policies.cc

```diff
@@ -746,6 +748,173 @@ class FailLbFactory : public LoadBalancingPolicyFactory {
   std::atomic<int>* pick_counter_;
 };
 
+//
+// QueueOnceLoadBalancingPolicy - a load balancing policy that provides a Queue
+// PickResult atleast once, after which it delegates to PickFirst.
+//
+
+constexpr char kQueueOncePolicyName[] = "queue_once";
+
+class QueueOnceLoadBalancingPolicy : public ForwardingLoadBalancingPolicy {
+ public:
+  QueueOnceLoadBalancingPolicy(Args args)
+      : ForwardingLoadBalancingPolicy(
+            std::make_unique<Helper>(
+                RefCountedPtr<QueueOnceLoadBalancingPolicy>(this)),
+            std::move(args), "pick_first",
+            /*initial_refcount=*/2) {}
+
+  ~QueueOnceLoadBalancingPolicy() override = default;
+
+  absl::string_view name() const override { return kQueueOncePolicyName; }
+
+ private:
+  class Helper;
+
+  // An always queueing picker
+  class QueueingPicker : public SubchannelPicker {
+   public:
+    QueueingPicker(Helper* helper) : helper_(helper) {}
+
+    PickResult Pick(PickArgs /*args*/) override;
+
+   private:
+    Helper* helper_;
+  };
+
+  class DelegatingPicker : public SubchannelPicker {
+   public:
+    DelegatingPicker(RefCountedPtr<SubchannelPicker> delegate_picker)
+        : delegate_picker_(std::move(delegate_picker)) {}
+
+    PickResult Pick(PickArgs args) override {
+      return delegate_picker_->Pick(args);
+    }
+
+   private:
+    RefCountedPtr<SubchannelPicker> delegate_picker_;
+  };
+
+  class Helper : public ChannelControlHelper {
+   public:
+    Helper(RefCountedPtr<QueueOnceLoadBalancingPolicy> parent)
+        : parent_(std::move(parent)) {}
+
+    RefCountedPtr<SubchannelInterface> CreateSubchannel(
+        ServerAddress address, const ChannelArgs& args) override {
+      return parent_->channel_control_helper()->CreateSubchannel(
+          std::move(address), args);
+    }
+
+    void UpdateState(grpc_connectivity_state state, const absl::Status& status,
+                     RefCountedPtr<SubchannelPicker> picker) override {
+      bool pick_queued;
```

Comment:
I think the logic here can be a bit simpler:

```
// If we've already seen a queued pick, just propagate the update directly.
if (parent_->SeenQueuedPick()) {
  parent_->channel_control_helper()->UpdateState(
      state, status, std::move(picker));
  return;
}
// Otherwise, store the update in the LB policy, to be propagated later, and return a queueing picker.
parent_->state_ = state;
parent_->status_ = status;
parent_->child_picker_ = std::move(picker);
parent_->channel_control_helper()->UpdateState(
    state, status, MakeRefCounted<QueuePicker>(parent_->Ref()));
```

The synchronization can happen inside of `SeenQueuedPick()`.

Reply:
Right, I didn't realize that the state could be synchronized through WorkSerializer by popping into it earlier from the picker, so we don't need to protect that with a mutex.